### PR TITLE
[JBTM-3163] AfterLRA notifications are delivered twice

### PR DESCRIPTION
**Pull request by xstefank**
https://issues.jboss.org/browse/JBTM-3163

No need to test. There is a TCK test in non-JAX-RS cases which is asserting that the afterLRA is called only once.

!QA_JTA !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS !PERF NO_WIN !AS_TESTS !TOMCAT !JACOCO !mysql !postgres !db2 !oracle !RTS

### CODE REVIEW

**Comment by xstefank** [comment]
@mmusgrov TckTests fails on different errors. Should I enable them then? Or do I need to implement changes required for them pass even if they are not required for this PR?

**Comment by mmusgrov** [comment]
They passed on our last CI run (Thursday): http://narayanaci1.eng.hst.ams2.redhat.com/job/narayana/PROFILE=LRA,jdk=jdk8.latest,label=linux/

I will push an update today that builds against the latest MP-LRA artifacts (if that is the issue).

**Comment by mmusgrov** [comment]
Also JBTM-3163 does not provide enough information.
Also there is nothing in the spec that says the notification cannot be sent more than once.
The PR or JIRA does not give any hint as to how the duplicates arise or how the PR stops duplicates being sent.

**Comment by xstefank** [comment]
I will update the JIRA description if it's not enough

**Comment by xstefank** [comment]
@mmusgrov I've updated the JIRA with my findings

File: rts/lra/lra-client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java [hunk 1]
```diff
@@ -677,7 +678,7 @@ private URI enlistCompensator(URI uri, long timelimit, String linkHeader, String
 
             if (response.getStatus() == Response.Status.PRECONDITION_FAILED.getStatusCode()) {
                 LRALogger.i18NLogger.error_tooLateToJoin(lraId, response);
-                throw new WebApplicationException(lraId + ": Too late to join with this LRA ", BAD_REQUEST);
+                throw new WebApplicationException(lraId + ": Too late to join with this LRA ", NOT_FOUND);
```
Comment (mmusgrov) [inline rts/lra/lra-client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java]: What is the motivation of this change?

File: rts/lra/lra-coordinator/src/main/java/io/narayana/lra/coordinator/domain/service/LRAService.java [hunk 2]
```diff
@@ -247,8 +247,6 @@ public LRAStatusHolder endLRA(URI lraId, boolean compensate, boolean fromHierarc
             }
         }
 
-        finished(transaction, fromHierarchy);
```
Comment (mmusgrov) [inline rts/lra/lra-coordinator/src/main/java/io/narayana/lra/coordinator/domain/service/LRAService.java]: Will this still work when the periodic recovery thread is running recovery

File: rts/lra/lra-test/tck/pom.xml [hunk 3]
```diff
@@ -43,6 +43,8 @@
                         <exclude>**/TckInvalidParticipantSignaturesTests.java</exclude>
                         <exclude>**/TckParticipantTests.java</exclude>
                         <exclude>**/TckInvalidSignaturesTests.java</exclude>
+                        <!-- https://github.com/eclipse/microprofile-lra/pull/192 -->
+                        <exclude>**/TckTests.java</exclude>
```
Comment (mmusgrov) [inline rts/lra/lra-test/tck/pom.xml]: We need to run the main test suite to validate that our implementation is working (we need to fix the TCK if it is wrong, not disable it). Is the issue that we need to update which TCK build artefact we are pulling in?

**Comment by mmusgrov** [comment]
> They passed on our last CI run (Thursday): http://narayanaci1.eng.hst.ams2.redhat.com/job/narayana/PROFILE=LRA,jdk=jdk8.latest,label=linux/
> 
> I will push an update today that builds against the latest MP-LRA artifacts (if that is the issue).

The tests still pass so this is not the issue. I will still take a look at updating the mp-lra snapshot dependency though, hopefully this week.

**Comment by xstefank** [comment]
I am already working on it to prove this is an issue.

**Comment by mmusgrov** [comment]
But do you agree we need to [change the test to assert that the notification](https://github.com/eclipse/microprofile-lra/blob/master/tck/src/main/java/org/eclipse/microprofile/lra/tck/TckTests.java#L259) was called at least once.

**Comment by xstefank** [comment]
Personally, I don't see any issue with delivering AfterLRA notification several times (but note that this is a different issue). Should I update that in the spec?

**Comment by mmusgrov** [comment]
The spec does not need to say how many times it should deliver the notification. We already say in the spec that messages can be lost.

But we do need a PR to change the test to assert that the notification is called at least once.

**Comment by xstefank** [comment]
> But we do need a PR to change the test to assert that the notification is delivered at least once.

yes, this is what I meant. We can change this in the specification. I will create a PR later today.

**Comment by xstefank** [comment]
Closing this as out of date. I will open a new PR when ready.

**Comment by mmusgrov** [comment]
@xstefank  I can submit a small fix if you haven't started yet

**Comment by xstefank** [comment]
@mmusgrov no I haven't. Go ahead please.
